File: othello_python/othello/board.py
"""
Class Board
Defines the game board
Akseli Lukkarila
2019-2023
"""

from colorprint import Color, get_color, print_color
from utils import Disk, Move, Square, Step


class Board:
    """Handles game board state and logic."""

    # Store all possible step directions on board as a class (static) variable
    STEP_DIRECTIONS = (
        Step(-1, -1),
        Step(-1, 0),
        Step(-1, 1),
        Step(0, -1),
        Step(0, 1),
        Step(1, -1),
        Step(1, 0),
        Step(1, 1),
    )

    def __init__(self, size=8):
        self._size = size
        # Initialize game board with empty disks.
        # Uses a 2D list ("list of rows") unlike other implementations.
        self._board = [[Disk.EMPTY for _ in range(self._size)] for _ in range(self._size)]

        # Set starting positions
        row = (self._size - 1) // 2 if self._size % 2 == 0 else (self._size - 1) // 2 - 1
        col = self._size // 2
        self._board[row][row] = Disk.BLACK
        self._board[row][col] = Disk.WHITE
        self._board[col][row] = Disk.WHITE
        self._board[col][col] = Disk.BLACK

        # Keep track of empty squares on board to avoid checking already filled positions
        self._empty_squares = {
            Square(x, y)
            for x in range(self._size)
            for y in range(self._size)
            if self._board[y][x] == Disk.EMPTY
        }

    def can_play(self) -> bool:
        """
        Return true if board contains empty squares.
        """
        return bool(self._empty_squares)

    def place_disk(self, player_move: Move) -> None:
        """Update board for given disk placement."""
        start = player_move.square
        assert (
            self._get_square(start) == Disk.EMPTY
        ), f"Trying to place disk to an occupied square {start}!"
        self._set_square(start, player_move.disk)
        self._empty_squares.remove(start)
        for step in player_move.directions:
            pos = start + step
            while self._get_square(pos) == player_move.disk.opponent():
                self._set_square(pos, player_move.disk)
                pos += step

    def possible_moves(self, disk: Disk) -> list[Move]:
        """Returns a list of all possible moves for the given disk colour."""
        moves = []
        opposing_disk = disk.opponent()
        for square in self._empty_squares:
            value = 0
            directions = []
            # try stepping in all directions from this starting point
            for step in self.STEP_DIRECTIONS:
                pos = square + step
                # Next square in this direction needs to be the opponents disk
                if self._get_square(pos) != opposing_disk:
                    continue
                steps = 0
                # Keep stepping over opponents disks
                while self._get_square(pos) == disk.opponent():
                    steps += 1
                    pos += step
                # Valid move only if a line of opponents disks ends in own disk
                if self._get_square(pos) == disk:
                    value += steps
                    directions.append(step)

            if value:
                moves.append(Move(square, disk, value, directions))

        return sorted(moves)

    def print_moves(self, moves: list[Move]) -> None:
        """Print available move coordinates and the resulting points gained."""
        print_color(f"  Possible moves ({len(moves)}):", Color.yellow)
        # Convert board from Disk enums to strings
        board = [[disk.board_char() for disk in row] for row in self._board]
        # Add possible moves to board
        for move in moves:
            x, y = move.square
            board[y][x] = get_color(str(move.value), Color.yellow)
            print(f"  {move}")

        # Print board with move positions
        print(f"    {' '.join(get_color(str(x), bold=True) for x in range(self._size))}")
        for index, row in enumerate(board):
            text = get_color(f"  {index} ", bold=True)
            text += " ".join(disk for disk in row)
            print(text)

    def print_score(self) -> None:
        """Print current score for both players."""
        black, white = self._player_scores()
        print(f"\n{self}")
        print(
            "Score: {} | {}".format(
                get_color(str(black), Disk.BLACK.color()), get_color(str(white), Disk.WHITE.color())
            )
        )

    def result(self) -> Disk:
        """Returns the winning disk colour. Empty indicates a draw."""
        total_score = self._score()
        if total_score == 0:
            return Disk.EMPTY

        return Disk.WHITE if total_score > 0 else Disk.BLACK

    def _check_coordinates(self, x: int, y: int) -> bool:
        """Check that the given coordinates are valid (inside the board)."""
        return 0 <= x < self._size and 0 <= y < self._size

    def _get_square(self, square: Square) -> Disk | None:
        """Returns the state of the board (empty, white, black) at the given square."""
        x, y = square
        return self._board[y][x] if self._check_coordinates(x, y) else None

    def _player_scores(self) -> tuple[int, int]:
        """Count and return the number of black and white disks."""
        white = sum(disk == Disk.WHITE for row in self._board for disk in row)
        black = sum(disk == Disk.BLACK for row in self._board for disk in row)
        return black, white

    def _score(self) -> int:
        """
        Returns the total score.
        Positive value means more white disks and negative means more black disks.
        """
        return sum(sum(row) for row in self._board)

    def _set_square(self, square: Square, disk: Disk) -> None:
        """Sets the given square to the given value."""
        x, y = square
        if not self._check_coordinates(x, y):
            raise ValueError(f"Invalid coordinates: {square}!")
        self._board[y][x] = disk

    def __str__(self) -> str:
        """Format game board to string."""
        text = f"  {' '.join(get_color(str(x), bold=True) for x in range(self._size))}"
        for index, row in enumerate(self._board):
            text += get_color(f"\n{index} ", bold=True)
            text += " ".join(disk.board_char() for disk in row)
        return text

File: cpp/src/board.cpp
//==========================================================
// Class Board source
// Defines the game board
// Akseli Lukkarila
// 2019-2026
//==========================================================

#include "board.hpp"

#include "colorprint.hpp"

#include <algorithm>  // std::ranges::sort, std::ranges::transform
#include <numeric>    // std::accumulate, std::iota
#include <ranges>     // std::ranges::transform (requires C++20)
#include <stdexcept>  // exceptions

namespace othello
{
/// Initialize a new board for the given board size.
Board::Board(const size_t size) :
    board(init_board(size)),
    // Keep track of empty squares on board to avoid checking already filled positions.
    empty_squares(init_empty_squares(size, board)),
    indices(size),
    size(size)
{
    // Index list (0...size) to avoid repeating same range in loops.
    std::iota(indices.begin(), indices.end(), 0);
}

/// Return true if board contains empty squares.
bool Board::can_play() const
{
    return !empty_squares.empty();
}

/// Update board for given disk placement.
void Board::place_disk(const Move& chosen_move)
{
    const auto start = chosen_move.square;
    const auto square = get_square(start);
    if (!square.has_value()) {
        throw std::invalid_argument(fmt::format("Invalid coordinates: {}", start));
    }
    if (square.value() != Disk::empty) {
        throw std::invalid_argument(
            fmt::format("Trying to place disk to an occupied square: {}!", start)
        );
    }
    set_square(start, chosen_move.disk);
    empty_squares.erase(start);
    for (const auto& affected_square : chosen_move.affected_squares()) {
        set_square(affected_square, chosen_move.disk);
    }
}

/// Returns a list of possible moves for the given player.
std::vector<Move> Board::possible_moves(Disk disk) const
{
    std::vector<Move> moves;
    const Disk opposing_disk = opponent(disk);
    for (const Square& square : empty_squares) {
        size_t value {0};
        std::vector<Direction> directions;
        for (const auto& step : STEP_DIRECTIONS) {
            Square pos {square + step};
            // Next square in this direction needs to be the opposing disk
            if (get_square(pos) != opposing_disk) {
                continue;
            }
            size_t num_steps {0};
            // Keep stepping over opponents disks
            while (get_square(pos) == opposing_disk) {
                ++num_steps;
                pos += step;
            }
            // Valid move only if a line of opposing disks ends with own disk
            if (get_square(pos) == disk) {
                directions.emplace_back(step, num_steps);
                value += num_steps;
            }
        }
        if (value > 0) {
            moves.emplace_back(square, disk, value, directions);
        }
    }
    std::ranges::sort(moves);
    return moves;
}

/// Print board with available move coordinates and the resulting points gained.
void Board::print_possible_moves(const std::vector<Move>& moves) const
{
    print_yellow("  Possible moves ({}):\n", moves.size());
    // Convert board from Disk enums to strings
    std::vector<std::string> formatted_board(board.size());
    std::ranges::transform(board, formatted_board.begin(), [&](const Disk& disk) {
        return board_char_with_color(disk);
    });
    // Add possible moves to board
    for (const Move& possible_move : moves) {
        const auto index = square_index(possible_move.square);
        formatted_board[index] = get_color(possible_move.value, fmt::terminal_color::yellow);
        fmt::print("  {}\n", possible_move);
    }
    // Print board with move positions
    fmt::print("    ");
    for (const auto i : indices) {
        print_bold(" {}", i);
    }
    for (const auto y : indices) {
        print_bold("\n  {}", y);
        for (const auto x : indices) {
            fmt::print(" {}", formatted_board[y * size + x]);
        }
    }
    fmt::print("\n");
}

/// Print current score for both players.
void Board::print_score() const
{
    const auto [black, white] = player_scores();
    fmt::print("\n{}\n", *this);
    fmt::print(
        "Score: {} | {}\n",
        get_color(std::to_string(black), disk_color(Disk::black)),
        get_color(std::to_string(white), disk_color(Disk::white))
    );
}

/// Returns the winning disk colour. Empty indicates a draw.
Disk Board::result() const
{
    using enum Disk;
    const int total_score = score();
    if (total_score == 0) {
        return empty;
    }
    return total_score > 0 ? white : black;
}

/// Get board status string for game log.
std::string Board::log_entry() const
{
    return std::accumulate(
        board.begin(),
        board.end(),
        std::string {},
        [](const std::string& accumulator, const Disk& disk) {
            return accumulator + board_char(disk);
        }
    );
}

/// Check that the given coordinates are valid (inside the board).
constexpr bool Board::check_coordinates(const int x, const int y) const
{
    return 0 <= x && x < static_cast<int>(size) && 0 <= y && y < static_cast<int>(size);
}

/// Check that the given square is valid (inside the board).
constexpr bool Board::check_square(const Square& square) const
{
    return check_coordinates(square.x, square.y);
}

/// Returns the state of the board (empty, white, black) at the given square.
std::optional<Disk> Board::get_square(const Square& square) const
{
    return check_square(square) ? std::optional {board[square_index(square)]} : std::nullopt;
}

/// Map square to board index.
constexpr size_t Board::square_index(const Square& square) const
{
    return static_cast<size_t>(square.y) * size + static_cast<size_t>(square.x);
}

/// Count and return the number of black and white disks.
std::tuple<int, int> Board::player_scores() const
{
    int black = 0;
    int white = 0;
    for (const auto& disk : board) {
        switch (disk) {
            case Disk::black:
                ++black;
                break;
            case Disk::white:
                ++white;
                break;
            default:
                break;
        }
    }
    return {black, white};
}

/// Returns the total score.
/// Positive value means more white disks and negative means more black disks.
int Board::score() const
{
    // Enum class prevents implicit conversion from Disk to int,
    // need to use lambda to cast Disk values to int.
    return std::accumulate(board.begin(), board.end(), 0, [](const int sum, Disk disk) {
        return sum + static_cast<int>(disk);
    });
}

/// Sets the given square to the given value.
void Board::set_square(const Square& square, const Disk disk)
{
    if (!check_square(square)) {
        throw std::invalid_argument(fmt::format("Invalid coordinates: {}", square));
    }
    board[square_index(square)] = disk;
}

/// Initialize game board with starting disk positions.
std::vector<Disk> Board::init_board(const size_t size)
{
    // Initialize game board with empty disks
    std::vector<Disk> board(size * size, Disk::empty);
    // Set starting positions
    const size_t row = size % 2 == 0 ? (size - 1) / 2 : (size - 1) / 2 - 1;
    const size_t col = size / 2;
    board[row * size + row] = Disk::white;
    board[row * size + col] = Disk::black;
    board[col * size + row] = Disk::black;
    board[col * size + col] = Disk::white;
    return board;
}

/// Initialize empty squares for the board.
std::set<Square> Board::init_empty_squares(const size_t size, const std::vector<Disk>& board)
{
    std::set<Square> empty_squares;
    for (size_t index = 0; index < board.size(); ++index) {
        if (board[index] == Disk::empty) {
            empty_squares.emplace(static_cast<int>(index % size), static_cast<int>(index / size));
        }
    }
    return empty_squares;
}

/// Format game board to string
std::ostream& operator<<(std::ostream& out, const Board& board)
{
    // Horizontal indices
    out << "  " << fmt::format(fmt::emphasis::bold, "{}", fmt::join(board.indices, " "));
    for (const auto& y : board.indices) {
        // Vertical index
        out << "\n" << fmt::format(fmt::emphasis::bold, "{}", y);
        // Row values
        for (const auto x : board.indices) {
            out << " " << board_char_with_color(board.board[y * board.size + x]);
        }
    }
    return out;
}
}  // namespace othello
